src/ristretto/bulletproofs_plus.rs

```diff
@@ -0,0 +1,629 @@
+                                "Generators' extension degree ({:?}) and proof's extension degree ({:?}) do not \
+                                 match; consider using a BulletproofsPlusService with a matching extension degree",
+                                self.generators.extension_degree(),
+                                rp.extension_degree()
+                            );
+                        }
+                        error!(target: LOG_TARGET, "Internal range proof error ({})", e.to_string());
+                        false
+                    },
+                }
+            },
+            Err(e) => {
+                error!(
+                    target: LOG_TARGET,
+                    "Range proof could not be deserialized ({})",
+                    e.to_string()
+                );
+                false
+            },
+        };
+    }
+
+    fn range(&self) -> usize {
+        self.generators.bit_length()
+    }
+}
+
+impl ExtendedRangeProofService for BulletproofsPlusService {
+    type K = RistrettoSecretKey;
+    type PK = RistrettoPublicKey;
+    type Proof = Vec<u8>;
+
+    fn construct_extended_proof(
+        &self,
+        masks: Vec<RistrettoExtendedMask>,
+        values: Vec<u64>,
+        min_value_promises: Vec<Option<u64>>,
+        seed_nonce: Option<Self::K>,
+    ) -> Result<Self::Proof, RangeProofError> {
+        if masks.is_empty() || values.is_empty() || masks.len() != values.len() {
+            return Err(RangeProofError::ProofConstructionError(
+                "Keys and values vectors cannot be empty and must have consistent length".to_string(),
+            ));
+        }
+        let mut commitments = Vec::with_capacity(values.len());
+        let mut openings = Vec::with_capacity(values.len());
+        for i in 0..values.len() {
+            // let extended_mask: Vec<Scalar> = masks[i].blindings()?.iter().map(|k| k.0).collect();
+            commitments.push(
+                self.generators
+                    .pc_gens()
+                    .commit(&Scalar::from(values[i]), &masks[i].as_scalar()?)
+                    .map_err(|e| RangeProofError::ProofConstructionError(e.to_string()))?,
+            );
+            openings.push(CommitmentOpening::new(values[i], masks[i].as_scalar()?));
+        }
+        let witness =
+            RangeWitness::init(openings).map_err(|e| RangeProofError::ProofConstructionError(e.to_string()))?;
+        let statement = RangeStatement::init(
+            self.generators.clone(),
+            commitments,
+            min_value_promises,
+            seed_nonce.map(|s| s.0),
+        )
+        .map_err(|e| RangeProofError::ProofConstructionError(e.to_string()))?;
+
+        let proof = RistrettoRangeProof::prove(self.transcript_label, &statement, &witness)
+            .map_err(|e| RangeProofError::ProofConstructionError(e.to_string()))?;
+
+        Ok(proof.to_bytes())
+    }
+
+    fn verify_batch_and_recover_masks(
```

Comment:
Batch recovery of masks is more expensive than linear mask recovery for the same amount of proofs, so that is not promoted. The primary action here is batch verification (logarithmic), with the additional benefit to recover masks at a linear cost. I will update the documentation to reflect this.